ocaml/database/master_connection.ml

```diff
@@ -215,6 +221,23 @@ let do_db_xml_rpc_persistent_with_reopen ~host:_ ~path (req : string) :
     else if !backoff_delay > 256.0 then
       backoff_delay := 256.0
   in
+  let rec delay = function
+    | d when d > 16.0 -> (
+        (* Sleep for a small time in case that the coordinator can connect but can't accept *)
+        delay (d /. 2.0) ;
+        match Atomic.get in_call_from_coordinator with
+        | true ->
+            debug
+              "%s: The coordinator is calling me. Wake up and retry connecting."
+              __FUNCTION__ ;
+            Atomic.set in_call_from_coordinator false
+        | false ->
+            delay (d /. 2.0)
```

Comment:
> Why wait when we are in call from the host that we try to connect to?

This is not in call from the coordinator, but it is holding the DB lock.
Another call from the coordinator (XAPI is restarting) would try to acquire the DB lock as well and hence be blocked.

> As written, we always delay d/2 first but skip the second delay d/2 when called from the coordinator. Is this the intention?

This first approach was to keep waking up from small timers (16 secs). Without the first delay, it might be connecting busily if the coordinator couldn't accept any connections yet.